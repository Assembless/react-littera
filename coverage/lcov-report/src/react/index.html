
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/react</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="..\..\prettify.css" />
    <link rel="stylesheet" href="..\..\base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(..\..\sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="..\..\index.html">All files</a> src/react</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">96.67% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>29/30</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">53.85% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>7/13</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">84.62% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>11/13</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">96.55% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>28/29</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
    </div>
    <div class='status-line high'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="hooks.ts"><a href="hooks.ts.html">hooks.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="25" class="abs high">25/25</td>
	<td data-value="85.71" class="pct high">85.71%</td>
	<td data-value="7" class="abs high">6/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="10" class="abs high">10/10</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="24" class="abs high">24/24</td>
	</tr>

<tr>
	<td class="file high" data-value="service.tsx"><a href="service.tsx.html">service.tsx</a></td>
	<td data-value="80" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 80%"></div><div class="cover-empty" style="width: 20%"></div></div>
	</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="5" class="abs high">4/5</td>
	<td data-value="16.67" class="pct low">16.67%</td>
	<td data-value="6" class="abs low">1/6</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="3" class="abs low">1/3</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="5" class="abs high">4/5</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank">istanbul</a>
                at Sat Sep 18 2021 13:07:55 GMT+0200 (Central European Summer Time)
            </div>
        </div>
        <script src="..\..\prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="..\..\sorter.js"></script>
        <script src="..\..\block-navigation.js"></script>
    </body>
</html>
    