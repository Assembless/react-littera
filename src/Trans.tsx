import * as React from "react";
import { TransProps } from "../types";

/**
 * This object contains the default HTML elements that can be parsed and rendered.
 */
const DEFAULT_ELEMENTS = {
  a: "a",
  b: "b",
  i: "i",
  strong: "strong",
  em: "em",
  p: "p",
  span: "span",
  div: "div",
  ul: "ul",
  ol: "ol",
  li: "li",
  h1: "h1",
  h2: "h2",
  h3: "h3",
  h4: "h4",
  h5: "h5",
  h6: "h6",
  br: "br",
};

/**
 * Trans is a custom translation component that can interpolate variables and parse HTML elements.
 *
 * @param {ReactNode} children - The translation string to be rendered.
 * @param {Record<string, string | ReactNode>} [values] - An object containing values to be interpolated in the translation string.
 * @param {Record<string, ElementType | string>} [components] - An object containing custom React components or HTML tags to be parsed and rendered.
 *
 * @returns {JSX.Element} A React element.
 *
 * @example
 *
 * ```jsx
 * import React from 'react';
 * import { Trans, useLittera } from '@assembless/react-littera';
 *
 * const translations = {
 *   welcome: {
 *     en_US: "Welcome <strong>{name}</strong>!",
 *     de_DE: "Willkommen <strong>{name}</strong>!",
 *     pl_PL: "Witaj <strong>{name}</strong>!",
 *   },
 * };
 *
 * function App() {
 *   const translated = useLittera(translations);
 *
 *   return (
 *     <Trans values={{ name: "Jack" }}>
 *       {translated}
 *     </Trans>
 *   );
 * }
 * ```
 */
const Trans: React.FC<TransProps> = ({ children, values = {}, components = {} }) => {
  /**
   * This function interpolates the given text string with the provided values.
   *
   * @param {string} text - The string to be interpolated.
   *
   * @returns {ReactNode} The interpolated string.
   */
  const interpolate = React.useCallback(
    (text: string): React.ReactNode => {
      let result = text;
      Object.entries(values).forEach(([key, value]) => {
        const pattern = new RegExp(`{${key}}`, "g");
        result = result.replace(pattern, String(value));
      });
      return result;
    },
    [values]
  );

  /**
   * This function takes a child node and returns a parsed React element.
   * If the child is a string, it will interpolate the string and parse any HTML tags found.
   * If the child is a React element, it will parse any child nodes it contains.
   *
   * @param {ReactNode} child - The child node to be parsed.
   *
   * @returns {ReactNode} The parsed child node.
   */
  const renderChild = React.useCallback(
    (child: React.ReactNode): React.ReactNode => {
      if (typeof child === "string") {
        const htmlTagRegex = /(<[^>]*>)/g; // Include the HTML tag in the regex match
        const parts = child.split(htmlTagRegex);

        return parts
          .map((part, index) => {
            if (htmlTagRegex.test(parts[index - 1] || "")) {
              // This part is an HTML tag, so convert it to a React component
              const tagName = parts[index - 1].replace(/<|>/g, "");
              const Component = components[tagName] || DEFAULT_ELEMENTS[tagName];
              if (Component) {
                return <Component key={index}>{interpolate(part)}</Component>;
              }
            }
            // This part is not an HTML tag, so interpolate it
            return interpolate(part);
          })
          .filter((part) => (typeof part === "string" ? !part.match(htmlTagRegex) : true));
      } else if (React.isValidElement(child)) {
        const { type, props } = child;
        const Component =
          typeof type === "string"
            ? components[type.toLowerCase()] ?? DEFAULT_ELEMENTS[type.toLowerCase()]
            : type;
        // Parse the child nodes.
        const childNodes = React.Children.map(props.children, renderChild);
        if (Component) {
          return <Component {...props}>{childNodes}</Component>;
        } else {
          const newProps = { ...props, children: childNodes };
          return React.cloneElement(child, newProps);
        }
      }
      return child;
    },
    [interpolate, components]
  );

  const renderedChildren = React.useMemo(
    () => React.Children.map(children, renderChild),
    [children, renderChild]
  );

  return <>{renderedChildren}</>;
};

export default Trans;
