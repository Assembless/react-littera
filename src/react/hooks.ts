import * as React from 'react'
import { LitteraContext } from './service'
import { translate } from '..'
import {
  deepMerge,
  throwInvalidLocale,
  warnMissingTranslations
} from '../utils/helpers'
import { LitteraTranslated, LitteraTranslations } from '../typings'

/**
 * Method accepting translations object and returning a React hook.
 * @param translations
 * @returns A React hook used to retrieve the translations.
 * @example
 * const translations = {
 *  'en_US': {
 *    'hello': 'Hello',
 *    'world': 'World'
 *  },
 *  'fr_FR': {
 *    'hello': 'Bonjour',
 *    'world': 'Monde'
 *  }
 * };
 * const useLittera = makeTranslations(translations);
 *
 * const Component () => {
 *  const translated = useLittera();
 *
 *  return <div>
 *    <h1>{translated.hello}</h1>
 *    <h2>{translated.world}</h2>
 *  </div>
 * }
 */
export const makeTranslations = <T, K extends keyof T>(
  translations: LitteraTranslations<T>
) => {
  warnMissingTranslations(translations)

  return (locale?: K) => {
    const service = React.useContext(LitteraContext)

    const translationsWithPreset = React.useMemo(
      () => deepMerge(service.preset, translations),
      [service.locale]
    ) as T & typeof service.preset

    return useLittera<T & typeof service.preset, K>(
      translationsWithPreset,
      locale
    )
  }
}

/**
 * @description React hook for translating a component.
 * @category React
 * @param translations
 * @param locale Optional locale to translate to. Defaults to the current locale.
 * @returns The translations object for the specified locale.
 */
export const useLittera = <T, K extends keyof T>(
  translations: LitteraTranslations<T>,
  locale?: K
): LitteraTranslated<T, K> => {
  const service = React.useContext(LitteraContext)
  const currentLocale: K = locale ?? service.locale

  React.useEffect(() => {
    if (locale) throwInvalidLocale(service.locales, locale as unknown as string)
  }, [locale])

  return React.useMemo(
    () => translate<T, K>(translations, currentLocale),
    [currentLocale]
  )
}

type LitteraMethodsReturn = {
  readonly locale: string
  readonly locales: string[]
  readonly setLocale: (locale: string) => void
  readonly translate: <T, K extends keyof T>(
    translations: LitteraTranslations<T>,
    locale?: string
  ) => LitteraTranslated<T, K>
}

/**
 * @description React hook exposing methods for current context.
 * @category React
 */
export const useLitteraMethods = () => {
  const { setLocale, locale, locales } = React.useContext(LitteraContext)

  const handleLocaleChange = React.useCallback(
    (nextLocale: string) => {
      throwInvalidLocale(locales, nextLocale)

      setLocale(nextLocale)
    },
    [setLocale]
  )

  const translateFn = <T, K extends keyof T>(
    translations: LitteraTranslations<T>,
    overrideLocale?: string
  ) => {
    const currentLocale = overrideLocale ?? locale ?? locales[0]

    return translate<T, K>(translations, currentLocale)
  }

  return React.useMemo<LitteraMethodsReturn>(
    () => ({
      setLocale: handleLocaleChange,
      locale,
      locales,
      translate: translateFn
    }),
    [locale]
  )
}
