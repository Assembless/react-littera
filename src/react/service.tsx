import * as React from 'react'
import { makeTranslations, useLitteraMethods, useLitteraRemote } from './hooks'
import { LitteraContextValue, LitteraRemoteOptions } from '../typings'
import { parseRemoteOptions } from '../utils/remote';
import { SWRConfig } from 'swr';

function localStorageProvider() {
  // When initializing, we restore the data from `localStorage` into a map.
  const map = new Map(JSON.parse(localStorage.getItem('app-cache') || '[]'))

  // Before unloading the app, we write back all the data into `localStorage`.
  window.addEventListener('beforeunload', () => {
    const appCache = JSON.stringify(Array.from(map.entries()))
    localStorage.setItem('app-cache', appCache)
  })

  // We still use the map for write & read for performance.
  return map
}

/**
 * @description Function initializing Littera. Creates a context and exposes Provider and other methods.
 * @param locales List of supported languages.
 * @param preset Global translations.
 * @returns The context, provider and hooks.
 */
export function createLittera<L extends ReadonlyArray<string>, P>(
  locales: L,
  preset: P,
) {
  const contextState: LitteraContextValue<L, P> = {
    locale: (locales[0] ?? 'en_US') as L[number],
    locales: locales as L,
    setLocale: () => { },
    preset,
    remote: undefined
  };

  const context = React.createContext(contextState);

  return {
    LitteraContext: context,
    /**
     * Context Provider for Littera
     * @category React
     * @public
     * @param initialLocale Initial active locale.
     * @param preset Set of predefined translations.
     * @param setLocale Callback called when the locale changes.
     * @param pattern Locale pattern.
     * @example
     * // Setting up Littera provider.
     *
     * const App = () => {
     *    return <LitteraService initialLocale={"en_US"}>
     *      ...
     *    </LitteraService>
     * }
     */
    LitteraService: ({
      children,
      initialLocale,
      remote
    }: {
      children: React.ReactNode
      initialLocale: L[number],
      remote?: LitteraRemoteOptions<P>,
    }) => {
      const [locale, setLocale] = React.useState(
        initialLocale ?? locales[0] ?? 'en_US'
      )
      const ContextProvider = context.Provider

      return (
        // @ts-ignore
        <SWRConfig value={{ provider: localStorageProvider }}>
          <ContextProvider
            value={{
              locale,
              setLocale,
              locales,
              preset,
              remote: parseRemoteOptions<P>(remote, locale)
            }}
          >
            {children}
          </ContextProvider>
        </SWRConfig>
      )
    },
    /**
     * Method accepting translations object and returning a React hook.
     * @param translations
     * @returns A React hook used to retrieve the translations.
     * @example
     * const translations = {
     *  'en_US': {
     *    'hello': 'Hello',
     *    'world': 'World'
     *  },
     *  'fr_FR': {
     *    'hello': 'Bonjour',
     *    'world': 'Monde'
     *  }
     * };
     * const useLittera = makeTranslations(translations);
     *
     * const Component () => {
     *  const translated = useLittera();
     *
     *  return <div>
     *    <h1>{translated.hello}</h1>
     *    <h2>{translated.world}</h2>
     *  </div>
     * }
     */
    makeTranslations:
      <T, Tp extends T & P, TpK extends keyof Tp>(translations: {
        [key in keyof T]: T[key]
      }) =>
        (
          locale?: TpK
        ): {
            [key in keyof Tp[TpK]]: Tp[TpK][key]
          } =>
          // @ts-ignore
          makeTranslations<L, P>(context)<Tp, TpK>(translations)(locale),
    /**
     * @description React hook exposing methods for current context.
     * @category React
     */
    useLitteraMethods: useLitteraMethods<L, P>(context),
    /**
     * @description React hook for translating a component from remote source.
     * @category React
     * @param path Path to the remote source.
     * @param locale Optional locale to translate to. Defaults to the current locale.
     * @returns The translations object for the specified locale.
     * @example
     * const Component = () => {
     *      const translated = useLitteraRemote("account/personal");
     * 
     *      return <div>
     *           <h1>{translated.firstName}</h1>
     *           <h2>{translated.age}</h2>
     *      </div>
     * }
     */
    useLitteraRemote: useLitteraRemote<L, P>(context)
  }
}
