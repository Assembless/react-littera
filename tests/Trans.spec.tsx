/**
 * @jest-environment jsdom
 */
import '@testing-library/jest-dom';
import * as React from 'react';
import { render, screen } from '@testing-library/react';
import Trans from '../src/Trans';

describe('<Trans />', () => {
  it('renders the translation string', () => {
    const { container } = render(<Trans>{"Welcome to our website!"}</Trans>);
    expect(screen.getByText('Welcome to our website!')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('interpolates the provided values', () => {
    const { container } = render(<Trans values={{ name: 'John' }}>{"Hello {name}!"}</Trans>);
    expect(screen.getByText('Hello John!')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('parses HTML tags in translation strings', () => {
    const { container } = render(<p><Trans>{"Hello <strong>World</strong>!"}</Trans></p>);
    expect(container.firstChild).toMatchSnapshot();
    expect(container.querySelector('strong')).toBeInTheDocument();
  });

  it('parses HTML tags from nodes', () => {
    const { container } = render(<p><Trans values={{ name: "Jack" }}>Hello <strong>{`{name}`}</strong>!</Trans></p>);
    expect(container.firstChild).toMatchSnapshot();
    expect(container.querySelector('strong')).toBeInTheDocument();
  });

  it('renders custom React components', () => {
    const CustomComponent = ({ children }) => <span data-testid="custom">{children}</span>;
    const { container } = render(
      <Trans components={{ custom: CustomComponent }}>
        {"<custom>Hello World!</custom>"}
      </Trans>
    );
    expect(screen.getByText('Hello World!')).toBeInTheDocument();
    expect(screen.getByTestId('custom')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
});