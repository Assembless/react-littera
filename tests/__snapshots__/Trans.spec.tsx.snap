// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Trans /> interpolates the provided values 1`] = `Hello John!`;

exports[`<Trans /> parses HTML tags 1`] = `
<p>
  Hello 
  <strong>
    World
  </strong>
  !
</p>
`;

exports[`<Trans /> renders custom React components 1`] = `
<span
  data-testid="custom"
>
  Hello World!
</span>
`;

exports[`<Trans /> renders the translation string 1`] = `Welcome to our website!`;
